clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java

```diff
@@ -136,11 +138,39 @@ public synchronized Sensor sensor(String name, MetricConfig config, Sensor... pa
         if (s == null) {
             s = new Sensor(this, name, parents, config == null ? this.config : config, time);
             this.sensors.put(name, s);
+            if (parents != null) {
+                for (Sensor parent : parents) {
+                    List<Sensor> children = childrenSensors.get(parent.name());
+                    if (children == null) {
+                        children = new ArrayList<>();
+                        childrenSensors.put(parent.name(), children);
+                    }
+                    children.add(s);
+                }
+            }
         }
         return s;
     }
 
     /**
+     * Remove a sensor (if it exists), associated metrics and its children.
+     *
+     * @param name The name of the sensor to be removed
+     */
+    public synchronized void removeSensor(String name) {
+        Sensor sensor = sensors.remove(name);
+        if (sensor != null) {
+            for (KafkaMetric metric : sensor.metrics())
+                removeMetric(metric.metricName());
+            List<Sensor> childSensors = childrenSensors.get(name);
+            if (childSensors != null) {
+                for (Sensor childSensor : childSensors)
+                    removeSensor(childSensor.name());
```

Comment:
Where do you cleanup the childrenSensors object? Otherwise we will maintain a reference to the Sensor objects always.